common/solver/batch_gmres_kernels.hpp.inc

```diff
@@ -0,0 +1,985 @@
+
+        if (r == 0) {
+            s_shared_entry.values[r * s_shared_entry.stride + c] =
+                z_norms_shared_entry.values[c];
+        }
+    }
+}
+
+
+template <typename BatchMatrixType_entry, typename ValueType, typename PrecType>
+__device__ __forceinline__ void arnoldi(
+    const BatchMatrixType_entry &A_global_entry, const int inner_iter,
+    const gko::batch_dense::BatchEntry<ValueType> &V_shared_entry,
+    const gko::batch_dense::BatchEntry<ValueType> &H_shared_entry,
+    const gko::batch_dense::BatchEntry<ValueType> &w_shared_entry,
+    const PrecType &prec_shared, const uint32 &converged)
+{
+    using real_type = typename gko::remove_complex<ValueType>;
+    const int &i = inner_iter;
+    const int num_rows = w_shared_entry.num_rows;
+    const int num_rhs = w_shared_entry.num_rhs;
+
+    const gko::batch_dense::BatchEntry<const ValueType> v_i_shared_entry{
+        &V_shared_entry.values[i * V_shared_entry.stride * num_rows],
+        V_shared_entry.stride, num_rows, V_shared_entry.num_rhs};
+
+    __shared__
+        UninitializedArray<ValueType, batch_config<ValueType>::max_num_rows *
+                                          batch_config<ValueType>::max_num_rhs>
+            w_temp_sh;
+
+    const gko::batch_dense::BatchEntry<ValueType> w_temp_shared_entry{
+        w_temp_sh, w_shared_entry.stride, w_shared_entry.num_rows,
+        w_shared_entry.num_rhs};
+
+    spmv_kernel(A_global_entry, v_i_shared_entry, w_temp_shared_entry);
+
+    __syncthreads();
+
+    prec_shared.apply(gko::batch::to_const(w_temp_shared_entry),
+                      w_shared_entry);
+    __syncthreads();
+
+
+    for (int k = 0; k <= i; k++) {
+        const gko::batch_dense::BatchEntry<const ValueType> v_k_shared_entry{
+            &V_shared_entry.values[k * V_shared_entry.stride * num_rows],
+            V_shared_entry.stride, num_rows, V_shared_entry.num_rhs};
+
+        const gko::batch_dense::BatchEntry<ValueType> h_k_i_shared_entry{
+            &H_shared_entry.values[k * H_shared_entry.stride + i * num_rhs],
+            static_cast<size_type>(w_shared_entry.num_rhs), 1,
+            w_shared_entry.num_rhs};
+
+        compute_dot_product(gko::batch::to_const(w_shared_entry),
+                            v_k_shared_entry, h_k_i_shared_entry);
+
+        __syncthreads();
+
+
+        for (int li = threadIdx.x;
+             li < w_shared_entry.num_rows * w_shared_entry.num_rhs;
+             li += blockDim.x) {
+            int r = li / w_shared_entry.num_rhs;
+
+            int c = li % w_shared_entry.num_rhs;
+
+            const uint32 conv = converged & (1 << c);
+
+            if (conv) {
+                continue;
+            }
+
+            ValueType h_k_i_scalar = h_k_i_shared_entry.values[c];
+
+            w_shared_entry.values[r * w_shared_entry.stride + c] -=
+                h_k_i_scalar *
+                v_k_shared_entry.values[r * v_k_shared_entry.stride + c];
+        }
+
+        __syncthreads();
+    }
```

Comment:
I am not sure how feasible it is, but maybe we can implement some form of multi-dot. Maybe that gives us some advantage in terms of lesser syncs due to lesser reductions. But it is also something we can look into later.

Reply:
This is not very clear to me. What is the exact meaning of multi-dot kernel? Here, there is a kind of dependency, value of w computed in every iteration depends upon the previous value of w.

Reply:
No, let's look into this later. It is just an idea, which would help in doing the different dot products at once.